File: bellows-ts/src/backends/postgres-discovery.ts
import { Pool } from "pg";
import {
  type PostgresBackendOptions,
  type PostgresDiscoveryCandidate,
  PostgresDiscoveryOperations,
  type PostgresSweepWindow,
} from "./postgres-operations.js";

export type {
  PostgresBackendOptions,
  PostgresDiscoveryCandidate,
  PostgresSweepWindow,
} from "./postgres-operations.js";

/**
 * Owned read-only PostgreSQL discovery without listeners, payload decoding, or claims.
 * Initialize schemas separately. The selected schema must belong entirely to the consuming
 * workload. On Workers use Hyperdrive with query caching disabled, never retain the backend across
 * events, and await close on success and error paths.
 */
export class PostgresDiscoveryBackend {
  private constructor(
    private readonly pool: Pool,
    private readonly operations: PostgresDiscoveryOperations,
  ) {}

  /** Validates/qualifies an optional existing schema; otherwise uses the default search path. */
  static async connect(
    databaseUrl: string,
    options: PostgresBackendOptions = {},
  ): Promise<PostgresDiscoveryBackend> {
    const pool = new Pool({ connectionString: databaseUrl });
    try {
      return new PostgresDiscoveryBackend(
        pool,
        new PostgresDiscoveryOperations(pool, options),
      );
    } catch (error) {
      await pool.end().catch(() => undefined);
      throw error;
    }
  }

  /** Captures database statement time and maximum task ID together, without a transaction. */
  async beginSweep(): Promise<PostgresSweepWindow> {
    return await this.operations.beginSweep();
  }

  /**
   * Reads at most 100 eligible identities in numeric ID order. Start with null, then use the
   * last returned ID even if consuming it failed; an empty page ends the pass. Page size bounds
   * query results, not consumer concurrency. Concurrent changes behind the cursor await another
   * pass; discovery does not reserve execution. Rows with null availability or an
   * availability at/before the cutoff are eligible, including expired leases with populated
   * owners. IDs retain the full signed-BIGINT range as text; never
   * round them through Number. Consumer-specific ID validation belongs after discovery.
   */
  async readPage(
    window: PostgresSweepWindow,
    lastSeenId: string | null,
  ): Promise<PostgresDiscoveryCandidate[]> {
    return await this.operations.readPage(window, lastSeenId);
  }

  /** Awaits owned pool shutdown. */
  async close(): Promise<void> {
    await this.pool.end();
  }
}

File: bellows/src/backends/postgres_discovery.rs
//! Read-only, listener-free PostgreSQL discovery. No payloads are decoded or tasks claimed.
//!
//! Enable `postgres` natively or `cloudflare` on wasm. Initialize schemas separately.

pub use super::postgres_common::{
    PostgresBackendOptions, PostgresDiscoveryCandidate, PostgresSweepWindow,
};
#[cfg(not(target_arch = "wasm32"))]
use super::postgres_operations::PostgresTaskOperations;
#[cfg(target_arch = "wasm32")]
use super::postgres_worker::PostgresTaskOperations;
#[cfg(target_arch = "wasm32")]
pub use super::postgres_worker::PostgresWorkerError;
#[cfg(target_arch = "wasm32")]
use PostgresWorkerError as DiscoveryError;
#[cfg(not(target_arch = "wasm32"))]
use sqlx::Error as DiscoveryError;

/// Owned discovery connections without listeners, schema initialization, or task definitions.
///
/// Clones share a native pool or request-scoped Workers connection. On Workers, use Hyperdrive
/// with query caching disabled and await [`Self::close`] on success and error paths; never retain
/// connections across events.
/// The selected schema must belong entirely to the workload consuming the discovered identities.
#[derive(Debug, Clone)]
pub struct PostgresDiscoveryBackend {
    operations: PostgresTaskOperations,
}

impl PostgresDiscoveryBackend {
    /// Uses the connection's default search path without changing it.
    pub async fn connect(database_url: &str) -> Result<Self, DiscoveryError> {
        Self::connect_with_options(database_url, PostgresBackendOptions::default()).await
    }

    /// Validates and qualifies an optional existing schema before connecting.
    pub async fn connect_with_options(
        database_url: &str,
        options: PostgresBackendOptions,
    ) -> Result<Self, DiscoveryError> {
        Ok(Self {
            operations: PostgresTaskOperations::connect(database_url, options).await?,
        })
    }

    /// Captures database statement time and the maximum task ID in one read.
    pub async fn begin_sweep(&self) -> Result<PostgresSweepWindow, DiscoveryError> {
        self.operations.begin_sweep().await
    }

    /// Reads at most 100 eligible identities in numeric ID order within the fixed window.
    ///
    /// Start with `None`, then pass the last returned ID, even if consuming that candidate failed.
    /// IDs retain the full signed-BIGINT range; consumer-specific ID validation belongs after discovery.
    /// An empty page ends the pass. This is not a snapshot: concurrent changes behind the cursor
    /// are left for the next pass. Eligibility does not reserve execution; claims remain authoritative.
    /// Rows are eligible when availability is null or at/before the cutoff, including
    /// expired leases with populated owners, for both published and singleton tasks.
    pub async fn read_page(
        &self,
        window: &PostgresSweepWindow,
        last_seen_id: Option<i64>,
    ) -> Result<Vec<PostgresDiscoveryCandidate>, DiscoveryError> {
        self.operations.read_page(window, last_seen_id).await
    }

    /// Awaits pool/driver shutdown, affecting all clones. Dropping a clone is insufficient.
    pub async fn close(&self) -> Result<(), DiscoveryError> {
        self.operations.close().await
    }
}
